src/core/derivation.ts

```diff
@@ -24,6 +29,30 @@ export interface IDerivation extends IDepTreeNode, IObservable, ISetEntry {
 	 * amount of dependencies used by the derivation in this run, which has not been bound yet.
 	 */
 	unboundDepsCount: number;
+	onBecomeStale();
+	recoverFromError();
+}
+
+export function shouldCompute(derivation: IDerivation, onError?): boolean {
+	const dependenciesState = derivation.dependenciesState;
+	if (dependenciesState === 0) return false;
+	if (dependenciesState === -1 || dependenciesState === 2) return true;
+	// derivation.dependencyChange === 1
+	let hasError = true;
+	try {
+		for (let i = 0; i < derivation.observing.length; i++) {
+			const obj = derivation.observing[i];
+			if (obj instanceof ComputedValue) {
+				obj.get();
```

Comment:
I see what this does, but calling `get()` and then relying on the side effects of `derivation.dependenciesState` might justify some elaborate comments :)